src/parquet/file/writer-internal.cc

```diff
@@ -174,9 +174,29 @@ int64_t SerializedPageWriter::WriteDictionaryPage(const DictionaryPage& page) {
 
 int RowGroupSerializer::num_columns() const { return metadata_->num_columns(); }
 
-int64_t RowGroupSerializer::num_rows() const { return num_rows_; }
+int64_t RowGroupSerializer::num_rows() const {
+  if (row_count_determined_) {
+    return num_rows_;
+  } else if (current_column_writer_) {
+    return current_column_writer_->RowsWritten();
+  } else {
+    return 0;
+  }
+}
 
 ColumnWriter* RowGroupSerializer::NextColumn() {
+  if (current_column_writer_) {
+   if (!row_count_determined_) {
+      num_rows_ = current_column_writer_->RowsWritten();
+     metadata_->SetNumRows(num_rows_);
```

Comment:
You should check if the same number of values are written across all columns in the arbitrary size case. You could make a short cut by just comparing with the first column written.